### DISCUSSION TITLE
CarPlay Maps with maui

### DISCUSSION THREAD

**Discussion by xvanneste** [dotnet/maui#17406] [Q&A]
Hello does anyone already create a CarPlay application with CPMapTemplate because I don't find any information about it just thé use of nscoder but I can't understand how to use it....
Thanks a lot

**Comment by mikelor** [comment]
Check out this series.
https://github.com/christian-strydom/MauiForCars

**Comment by ghost** [comment]
This is for GridTemplate, but maybe somebody can benefit from this process we went trough.

**1. Requesting CarPlay Entitlements**
Read more info on official Apple website.
https://developer.apple.com/documentation/carplay/requesting-carplay-entitlements

**2. Add the entitlement into Entitlements.plist**
Make sure you add the entitlement you have been granted.
Maybe you have asked for X and you got Y. It won't work if the entitlement will not match with your provisiong profiles.

```
<key>com.apple.developer.carplay-driving-task</key>
<true/>	
```

**3. Visual Studio - updating profiles**
Magic to me. In the end I added Apple account again into Visual Studio and clicked "Download all profiles"

**4. Add information to info.plist**
```
<key>UIApplicationSceneManifest</key>
	<dict>
		<key>UIApplicationSupportsMultipleScenes</key>
		<true/>
		<key>UISceneConfigurations</key>
		<dict>
			<!-- Default MAUI scene for iOs -->
			<key>UIWindowSceneSessionRoleApplication</key>
			<array>
				<dict>
					<key>UISceneClassName</key>
					<string>UIWindowScene</string>
					<key>UISceneConfigurationName</key>
					<string>__MAUI_DEFAULT_SCENE_CONFIGURATION__</string>
					<key>UISceneDelegateClassName</key>
                    <!-- iOs scene, match "SceneDelegate" with your class name, other fields can be left as they are -->
					<string>SceneDelegate</string>
				</dict>
			</array>
			<!-- CarPlay scene, match "CarPlaySceneDelegate" with your class name, other fields can be left as they are -->
			<key>CPTemplateApplicationSceneSessionRoleApplication</key>
			<array>
				<dict>
					<key>UISceneClassName</key>
					<string>CPTemplateApplicationScene</string>
					<key>UISceneConfigurationName</key>
					<string>CarPlay</string>
					<key>UISceneDelegateClassName</key>
					<string>CarPlaySceneDelegate</string>
				</dict>
			</array>
		</dict>
	</dict>
```
	
**5. MauiUISceneDelegate.cs in Platforms/iOS/ - iOS**
```
using Foundation;

namespace Maui.Platforms.iOS
{
    [Register("SceneDelegate")]
    public class SceneDelegate : MauiUISceneDelegate
    {
        
    }
}
```

**6. CarPlaySceneDelegate.cs in Platforms/iOS - CarPlay**
```
using CarPlay;
using CommunityToolkit.Mvvm.Messaging;
using Foundation;

namespace Maui.Platforms.iOS
{
    [Register("CarPlaySceneDelegate")]
    public class CarPlaySceneDelegate : CPTemplateApplicationSceneDelegate
    {
        private CPGridTemplate? cPGridTemplate = null;
        private List<CPGridButton> gridButtons = [];
        private bool dataReload = false;

        private IDispatcherTimer? _reloadTimeoutTimer;

        public CarPlaySceneDelegate()
        {
            // When data changes, we can call from anywhere in the MAUI to reload the data also on the CarPlay screen
            WeakReferenceMessenger.Default.Unregister<CarReloadDataMessage>(this);
            WeakReferenceMessenger.Default.Register<CarReloadDataMessage>(this, (recipient, message) =>
            {
                dataReload = true;
            });

            // Is timer up from previous session?
            if (_reloadTimeoutTimer != null)
            {
                _reloadTimeoutTimer?.Stop();
                _reloadTimeoutTimer = null;
            }

            // Start with new timer for refreshing grid buttons
            _reloadTimeoutTimer = Dispatcher.GetForCurrentThread()?.CreateTimer();
            if (_reloadTimeoutTimer != null)
            {
                _reloadTimeoutTimer.Interval = TimeSpan.FromMilliseconds(250);
                _reloadTimeoutTimer.Tick += async (s, e) =>
                {
                    if (dataReload)
                    {
                        await MainThread.InvokeOnMainThreadAsync(() =>
                        {
                            GetCpGridButtons(false);
                            cPGridTemplate?.UpdateGridButtons([.. gridButtons]);
                        });

                        dataReload = false;
                    }
                };

                _reloadTimeoutTimer.IsRepeating = true;
                _reloadTimeoutTimer.Start();
            }
        }

        private void GetCpGridButtons(bool firstLoad)
        {
            try
            {
                // Get list of Grid buttons
                int limit = 0;
                gridButtons = [];

                // Get all data
                foreach (Data tempData in AllData)
                {
                    // We can only show up to 8 controls on grid list
                    if (limit >= (int)CPGridTemplate.MaximumItems)
                    {
                        limit++;
                        break;
                    }


                    // Add a button with on click handler
                    gridButtons.Add(
                      new CPGridButton(
	                      titleVariants: [tempData.Name],
	                      UIKit.UIImage.FromBundle(tempData.Image),
	                      (gridButton) =>
	                      {
		                      // Do smt on click, for example light on
		                      // In 250 ms we will refresh and the Grid button will get the yellow light image
	                      }
                      )
                    );

                    limit++;
                }
            }
            catch (Exception ex)
            {
                Console.Writeline("Reloading CarPlay view failed");
            }
        }

        [Export("templateApplicationScene:didConnectInterfaceController:to:")]
        public override void DidConnect(CPTemplateApplicationScene templateApplicationScene, CPInterfaceController interfaceController)
        {
            GetCpGridButtons(true);
            cPGridTemplate = new("Maui", [.. gridButtons]);
            interfaceController.SetRootTemplate(cPGridTemplate, false);
        }

        [Export("templateApplicationScene:didDisconnectInterfaceController:from:")]
        public override void DidDisconnect(CPTemplateApplicationScene cpTemplateApplicationScene, CPInterfaceController interfaceController)
        {
            // Clear the timer
            if (_reloadTimeoutTimer != null)
            {
                _reloadTimeoutTimer?.Stop();
                _reloadTimeoutTimer = null;
            }

            interfaceController.Dispose();
        }
    }
}
```

**References - code above was created with help of links bellow and [CoPilot AI](https://copilot.microsoft.com/):**
 - [Maui - info.plist info](https://stackoverflow.com/questions/56093637/is-it-possible-to-develop-a-carplay-app-using-xamarin-tools)
 - [Maui - scene delegate](https://learn.microsoft.com/en-sg/answers/questions/1498516/blank-screen-after-adding-carplay-to-maui-app)
 - [Swiftui - can help with Maui understanding](https://stackoverflow.com/questions/76198577/ios-swiftui-carplay-hello-world)
 - [Where is CarPlay available?](https://www.apple.com/ios/carplay/available-models/)
 - [I notice UIApplicationSupportsMultipleScenes - key to get is started](https://stackoverflow.com/questions/77050489/xamarin-carplay-supported-ios-app-crashes-when-starting)
 - [Was looking for lifecycle, how to refresh CarPlay data?](https://developer.apple.com/documentation/technotes/tn3187-migrating-to-the-uikit-scene-based-life-cycle)
 - [Good overall knowledge](https://www.fokswang.com/post/47-compatible-with-uiscene)
 - [didConnect handler](https://developer.apple.com/documentation/carplay/cptemplateapplicationscenedelegate/templateapplicationscene(_:didconnect:))
 - [Is UIApplicationSupportsMultipleScenes missing?](https://stackoverflow.com/questions/79077404/net-maui-apple-carplay-black-screen)
 - [Searching for UISceneConfigurationName help](https://duckduckgo.com/?q=%3Ckey%3EUISceneConfigurationName%3C%2Fkey%3E+%2B+maui+%2B+carplay&atb=v436-1&ia=web)
 - [How to know when CarPlay is connected?](https://duckduckgo.com/?q=CarPlaySceneDelegate.IsCarPlayConnected&atb=v436-1&ia=web)
 - [Overall knowledge, also for MAUI](https://stackoverflow.com/questions/61539391/xamarin-forms-for-ios-using-scenedelegate-cs/74315179#74315179)
 - [Good code example for CarPlay](https://github.com/christian-strydom/MauiForCars/blob/main/MauiForCars/MauiForCars/Platforms/iOS/CarPlaySceneDelegate.cs)
 - [Good code example for CarPlay - 2](https://github.com/christian-strydom/MauiForCars/blob/main/MauiForCars/MauiForCars/Platforms/iOS/DeviceSceneDelegate.cs)
 - [Duck search for starting](https://duckduckgo.com/?q=maui+carplay&atb=v436-1&ia=web)
 - [Reddit post - good video](https://www.reddit.com/r/dotnetMAUI/comments/13cjz9f/net_maui_for_cars_apple_carplay_tutorial/)
 - [Official CarPlay Developer Guide](https://developer.apple.com/download/files/CarPlay-Developer-Guide.pdf)
 - [Official CarPlay - CPGridTemplate info](https://developer.apple.com/documentation/carplay/cpgridtemplate)

**Reply by davidortinau** [reply]
@rokmeglic71 I'm curious about your experience building for CarPlay, and what your use case is. If you're interested in a chat, shoot me an email david.ortinau@microsoft.com

**Reply by development-jantar** [reply]
The new Luqid design brakes the CarPlay Grid Template Screen.
<img width="1680" height="1124" alt="unnamed" src="https://github.com/user-attachments/assets/5917afdc-a469-445b-ba92-ad7294bc9d5f" />

**Reply by development-jantar** [reply]
We figured out. If you create a new GridTemplate with empty buttons and later do the                             `cPGridTemplate?.UpdateGridButtons([.. gridButtons]);,` all the buttons will have position 0,0 (top left corner).

To fix this, create buttons first, so first call to new cpGridTemplate will have some buttons inside.

```
            // Create list of buttons
            GetCpGridButtons();

            // cpGridTemplate
            cPGridTemplate = new(StatusText, [.. gridButtons]);
            interfaceController.SetRootTemplate(cPGridTemplate, false);
```
**CoPilot - possible explanation:**

> The issue happened because the CPGridTemplate was created with zero buttons.
> CarPlay calculates the layout at creation time, so when the template starts empty, the layout collapses to position (0,0).
> Later updates to the button list do not trigger a new layout pass, so all buttons stay stuck in the top‑left corner.

**Reply by development-jantar** [reply]
Unfortountly latest xCode 26.0 (or maybe was this present also in previous versions) do not work well with `cPGridTemplate?.UpdateGridButtons([.. gridButtons]);`. We are experiencing that after phone reconnects on second session, the grid is empty or all buttons are positioned on top left corner.

So instead using `cPGridTemplate?.UpdateGridButtons([.. gridButtons]);` we create new template and set it as root template.
```
cPGridTemplate = new(StatusText, [.. gridButtons]);
interfaceController?.SetRootTemplate(cPGridTemplate, false);
```

**Reply by rokmeglicbit** [reply]
[There is another CarPlay & Android guide available now from Allan Ritchie](https://allanritchie.com/blog/2026/02/maui-carplay/) and Pull Request for [[net12.0][iOS] Add CarPlay, WidgetKit support and scene delegate fixes](https://github.com/dotnet/maui/pull/34234)
